algorithms/kernel/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i

```diff
@@ -71,289 +92,249 @@ Status KNNClassificationPredictKernelUCAPI<algorithmFpType>::compute(const Numer
+            BlockOfRows curDataRows(dblock, maxDataBlockRows, nDataRows);
+            BlockDescriptor<int> labelRows;
+            DAAL_CHECK_STATUS_VAR(labels->getBlockOfRows(curDataRows.first, curDataRows.number, readOnly, labelRows));
+            auto dataLabels = labelRows.getBuffer();
+            BlockDescriptor<algorithmFpType> dataRows;
+            DAAL_CHECK_STATUS_VAR(points->getBlockOfRows(curDataRows.first, curDataRows.number, readOnly, dataRows));
+            auto curData = dataRows.getBuffer();
+            // Collect sums of squares from train data
+            auto sumResult = math::SumReducer::sum(math::Layout::RowMajor, curData, curDataRows.number, nFeatures, &st);
             DAAL_CHECK_STATUS_VAR(st);
-            initDistances(context, init_distances, sumResult.sumOfSquares, distances, curDataBlockSize, curProbeBlockSize, &st);
+            // Initialize GEMM distances
+            scatterSumOfSquares(context, sumResult.sumOfSquares, distances, curDataRows.number, curQueryRows.number, &st);
             DAAL_CHECK_STATUS_VAR(st);
-            computeDistances(context, data, probes, distances, curDataBlockSize, curProbeBlockSize, nFeatures, &st);
+            // Let's calculate distances using GEMM
+            computeDistances(context, curData, curQuery, distances, curDataRows.number, curQueryRows.number, nFeatures, &st);
             DAAL_CHECK_STATUS_VAR(st);
-            {
-                DAAL_ITTNOTIFY_SCOPED_TASK(compute.copy_categories);
-                initCategories(context, init_categories, labels_all, categories, curProbeBlockSize, curDataBlockSize, dfirst, &st);
-                DAAL_CHECK_STATUS_VAR(st);
-            }
-            QuickSelectIndexed::select(distances, categories, rndSeq, dataBlockSize, nK, curProbeBlockSize, curDataBlockSize, curDataBlockSize,
-                                       selectResult, &st);
+            // Select nK smallest distances and their labels from every row of the [curQueryRows.number]x[curDataRows.number] block
+            selector->selectNearestDistancesAndLabels(distances, dataLabels, nK, curQueryRows.number, curDataRows.number, curDataRows.number, 0,
+                                                      selectResult, &st);
             DAAL_CHECK_STATUS_VAR(st);
-            copyPartialSelections(context, gather_selection, selectResult.values, selectResult.indices, partialDistances, partialCategories,
-                                  curProbeBlockSize, nK, partCount, nParts, dblock != 0 && partCount == 0, &st);
+            // copy block results to buffer in order to get merged with the same selection algorithm (up to selectionNumberOfChunks of partial results)
+            // and keep the first part containing previously merged result if exists
+            copyPartialSelections(context, selectResult.values, selectResult.indices, partialDistances, partialLabels, curQueryRows.number, nK,
```

Comment:
For all your functions, please arrange your parameters in the next order:
input1, ..., input_k, output1, ..., output_n